src/Credentials/CachingProviderTrait.php

```diff
@@ -0,0 +1,67 @@
+<?php
+
+namespace Aws\Credentials;
+
+use Doctrine\Common\Cache\Cache;
+use Doctrine\Common\Cache\PhpFileCache;
+
+trait CachingProviderTrait
+{
+    private static $configCredentialsCache = 'credentials.cache';
+    private static $configCredentialsCacheKey = 'credentials.cache.key';
+
+    private static function tryToCache(
+        CredentialsInterface $credentials,
+        array $config = []
+    ) {
+        $cache = self::getCredentialsCache($config);
+        if (!$cache || $credentials->isExpired()) {
+            return null;
+        }
+
+        $cache->save(
+            self::getCacheKey($config),
+            $credentials,
+            $credentials->getExpiration() - time()
+        );
+    }
+
+    /**
+     * @param array $config
+     * @return Cache|null
+     */
+    private static function getCredentialsCache(array $config = [])
+    {
+        if (!empty($config[self::$configCredentialsCache])) {
+            return $config[self::$configCredentialsCache] instanceof Cache ?
+                $config[self::$configCredentialsCache]
+                : self::getDefaultCache();
+        }
+
+        return null;
+    }
+
+    private static function getDefaultCache()
+    {
+        $cacheDir = sys_get_temp_dir() .
```

Comment:
Don't we already have a default location for AWS caching that the JSON loader uses?